File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        int n = nums.size();
        unordered_map<int, int> counts;
        for (int num : nums) {
            counts[num]++;
        }
        vector<vector<int>> buckets(n + 1);
        for (auto& [num, freq] : counts) {
            buckets[freq].push_back(num);
        }
        vector<int> result;
        result.reserve(k);         
        for (int freq = n; freq >= 0 && result.size() < k; freq--) {
            if (!buckets[freq].empty()) {
                for (int num : buckets[freq]) {
                    result.push_back(num);
                    if (result.size() == k) break;
                }
            }
        }
        return result;
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {

        HashMap<Integer, Integer> frequency = new HashMap<>();

        for (int num : nums) {
            frequency.put(num, frequency.getOrDefault(num, 0) + 1);
        }

        PriorityQueue<Integer> minHeap =
            new PriorityQueue<>((a, b) ->
                frequency.get(a) - frequency.get(b));

        for (int num : frequency.keySet()) {
            minHeap.offer(num);

            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }

        int[] result = new int[k];

        for (int i = k - 1; i >= 0; i--) {
            result[i] = minHeap.poll();
        }

        return result;
    }
}